extensions/eclipse-che-theia-plugin-ext/src/common/che-protocol.ts

```diff
@@ -0,0 +1,174 @@
+/*********************************************************************
+ * Copyright (c) 2018 Red Hat, Inc.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ **********************************************************************/
+
+import { ProxyIdentifier, createProxyIdentifier } from '@theia/plugin-ext/lib/api/rpc-protocol';
+
+export interface CheApiPlugin {
+
+}
+
+export interface CheApiMain {
+    $currentWorkspace(): Promise<Workspace>;
+
+    $getFactoryById(id: string): Promise<Factory>;
+}
+
+export interface Factory {
+    workspace: WorkspaceConfig;
+    ide?: {
```

Comment:
Does it related to CHE IDE or also to Theia?

Reply:
It's just factory configuration https://www.eclipse.org/che/docs/che-6/factories_json_reference.html

Reply:
Che is able to produce typescript definition of current DTOs
https://github.com/eclipse/che/blob/21318aeec878a02ee9a64453ad8e737204ccfe21/dockerfiles/lib/build.sh#L35-L44

might be good if we could reuse these definitions automatically instead of trying to duplicate it

Reply:
Good catch, I will look at this

Reply:
as an intermediate step, we could just move out `upstream che definitions` to a separate .d.ts file for now